Check if role member has been disabled during authz checks

patrasap0908 opened the pull request:
This PR performs an additional check to deny access for role members that have been disabled on Athenz, as described by https://github.com/yahoo/k8s-athenz-webhook/issues/22.

<!-- The following line must be included in your pull request -->
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

RuiGuo34 commented:
@mcieplak fyi: go client version is pretty old, need to update to 1.17 sometime later

RuiGuo34 commented:
@patrasap0908 Can you add a simple test case regarding systemdisabled check?

mcieplak commented on zpeclient.go:
```diff
@@ -328,7 +332,7 @@ func (c *Cache) authorize(principal string, check AthenzAccessCheck) (bool, erro
 		for _, member := range members {
 			if member.memberRegex.MatchString(principal) {
-				if member.expiration.IsZero() || member.expiration.After(time.Now()) {
+				if member.expiration.IsZero() || member.expiration.After(time.Now()) && !member.systemDisabled {
```
Shouldn't we put the `member.expiration.IsZero() || member.expiration.After(time.Now())` check into parenthesis? Otherwise, the first `member.expiration.IsZero()` can short circuit if it's true and then the systemDisabled flag will be essentially ignored.

RuiGuo34 replied:
+1 on that

patrasap0908 replied:
Oops, my bad. Should be good now - https://github.com/yahoo/k8s-athenz-webhook/pull/24/files#diff-abac4b8aebeabb41b0d0f0baa1d6ec2f2c0920708fa80c5282be51c4c1ff412bR335.

mcieplak commented:
@patrasap0908 +1 for the unit tests. Can we add some simple ones similar to the one we have for an expired user [here](https://github.com/yahoo/k8s-athenz-webhook/blob/master/zpeclient_test.go#L713-L739).

patrasap0908 commented:
Yep, added a test for this - https://github.com/yahoo/k8s-athenz-webhook/pull/24/files#diff-93c4dfb69ea53188fa218a8bf3d778236499b823f6a6911c773f78e6dc919226R741-R770.

Made a couple of small changes:
- Added a `break` statement to the loop while processing role members so that further iterations are skipped once principal match has happened successfully - https://github.com/yahoo/k8s-athenz-webhook/pull/24/files#diff-abac4b8aebeabb41b0d0f0baa1d6ec2f2c0920708fa80c5282be51c4c1ff412bR337.
- Fixed the test for member expiration checks. It was for the resource `home.domain:pods` but the new role was getting created in `home.domain.deny` domain as a result of which, the `authorize()` method always returned `false` (which is the default behaviour when no roles are found for the passed principal) - https://github.com/yahoo/k8s-athenz-webhook/pull/24/files#diff-93c4dfb69ea53188fa218a8bf3d778236499b823f6a6911c773f78e6dc919226L716-R716.

mcieplak commented on zpeclient.go:
```diff
@@ -328,8 +332,9 @@ func (c *Cache) authorize(principal string, check AthenzAccessCheck) (bool, erro
-				if member.expiration.IsZero() || member.expiration.After(time.Now()) {
+				if (member.expiration.IsZero() || member.expiration.After(time.Now())) && !member.systemDisabled {
 					roles = append(roles, role)
+					break
```
We can't do a break here because the member can be part of multiple different roles and only one of them might have the permissions to give access. With this change, users will actually be denied unless that first role has the policy to allow them.

patrasap0908 replied:
My bad, you're right. Removed the `break` - https://github.com/yahoo/k8s-athenz-webhook/pull/24/commits/861591ef716779739bee91912b67c44ceb37ab39.
